src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -817,12 +830,7 @@ public override void VisitArraySyntax(ArraySyntax syntax)
                     return ErrorType.Create(errors);
                 }
 
-                if (TypeHelper.TryCollapseTypes(itemTypes) is not { } collapsedItemType)
-                {
-                    return LanguageConstants.Array;
-                }
-
-                return new TypedArrayType(collapsedItemType, TypeSymbolValidationFlags.Default);
+                return new TupleType(typeName.ToString(), itemTypes.ToImmutableArray<ITypeReference>(), TypeSymbolValidationFlags.Default);
```

Comment:
I'm assuming the reason we don't have the `TupleType` constructor generate the name internally is that we aren't always able to evaluate the deferred type references without causing a cycle. What do we do for naming in those cases?

Reply:
Cycles should be fine; the namer uses `SyntaxBase::ToText` instead of calling `.Type.Name` if it encounters a deferred type reference. The name is currently generated externally just to avoid looping over the tuple children twice. It's slightly more efficient but makes for a clunkier constructor signature.